doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1257,157 +1257,155 @@ Managing the local context
 
-      .. coqtop:: reset all
+      .. coqtop:: in
 
-         Goal forall x :nat, x = 0 -> forall z y:nat, y=y-> 0=x.
-           intros x H z y H0.
-           move x after H0.
+           (*                    x Hx y z Hy *)
+           move y after z.  (*    x Hx z y Hy   (z was left of y, intuitive case) *)
            Undo.
-           move x before H0.
+           move z after y.  (*    x Hx z y Hy   (z was right of y, see Note above) *)
            Undo.
-           move H0 after H.
+           move x after Hy.  (*   y z Hy x Hx   (Hx depends on x, so moved) *)
            Undo.
-           move H0 before H.
+           move x before Hy.  (*  y z x Hx Hy *)
+           Undo.
+           move Hy after Hx.  (*  x y Hy Hx z *)
+           Undo.
+           move Hy before Hx.  (* x Hx y Hy z *)
 
-.. tacn:: rename @ident__1 into @ident__2
-   :name: rename
+.. tacn:: rename {+, @ident__1 into @ident__2 }
 
-   This renames hypothesis :n:`@ident__1` into :n:`@ident__2` in the current
-   context. The name of the hypothesis in the proof-term, however, is left
-   unchanged.
+   Renames hypothesis :n:`@ident__1` into :n:`@ident__2` for each pair of :n:`@ident`\s.
+   Renaming is done simultaneously, which permits swapping the names of 2 hypotheses.
+   (Note that the renaming is applied in the context and the existential
+   variables, but the proof term doesn't change.)
 
-   .. tacv:: rename {+, @ident__i into @ident__j}
+   .. exn:: @ident is already used.
+      :undocumented:
 
-      This renames the variables :n:`@ident__i` into :n:`@ident__j` in parallel.
-      In particular, the target identifiers may contain identifiers that exist in
-      the source context, as long as the latter are also renamed by the same
-      tactic.
+.. tacn:: set @one_term {? @as_name } {? @occurrences }
+          set @bindings_with_parameters {? @occurrences }
+   :name: set; _
 
-   .. exn:: No such hypothesis.
-      :undocumented:
+   .. insertprodn bindings_with_parameters bindings_with_parameters
 
-   .. exn:: @ident is already used.
-      :undocumented:
+   .. prodn::
+      bindings_with_parameters ::= ( @ident {* @simple_binder } := @term )
 
-.. tacn:: set (@ident := @term)
-   :name: set
+      .. todo PR: Should bindings_with_parameters be
+            just binding_with_parameters (singular "binding")?
+            Or maybe call it "function_binding"??
+         Also, not much description of "simple_binder", which appears in tacn:fix
 
-   This replaces :token:`term` by :token:`ident` in the conclusion of the
-   current goal and adds the new definition :n:`@ident := @term` to the
-   local context.
+   The first form replaces :n:`@one_term` in the goal with a fresh variable
+   and adds the variable as a new :term:`local definition <context-local definition>`.
+   Use :n:`@as_name` to specify the name of the new variable.
+
+   The second form is equivalent to
+   :n:`set (@ident := fun {* @simple_binder } => @one_term) …`.
```

Comment:
This is a circular definition: a special case of the second form (with `simple_binder`) is defined in terms of another case of the second form (without `simple_binder`), without the latter being defined at all.

BTW, the `one_term` should be `term` here.